### ISSUE TITLE
RPC ListValidatorBalances index out range error

### ISSUE THREAD

**Issue by imnisen** [prysmaticlabs/prysm#7838]
https://github.com/prysmaticlabs/prysm/blob/ad5151f25d0aa46e1357eb56dc5079d7d32a725a/beacon-chain/rpc/beacon/validators.go#L30

Hi, when requesting `ListValidatorBalances` api with a page_size and a non-existed pubkey (such as eth/v1alpha1/validators/balances?epoch=1346&pageSize=100&publicKeys=xxxx) , this handler would crash due to index out of range error. Which occured in my beacon node.

After reading the code, I find that if pass the non-existed pubkey, `res` would contain one item while `balancesCount` is the length total validators, then `end` is also the total length, then `res[start:end]` would cause out of range index error.

https://github.com/prysmaticlabs/prysm/blob/ad5151f25d0aa46e1357eb56dc5079d7d32a725a/beacon-chain/rpc/beacon/validators.go#L75
It seems not nessary to set `balancesCount` to total length first: `balancesCount := len(balances)`. As it is only needed in `if len(req.Indices) == 0 && len(req.PublicKeys) == 0 `situation , which could caculated alone.

Part of the error logs:

```
[[90m[2020-11-17 16:52:52]^[[0m ^[[31mERROR^[[0m gRPC panicked! ^[[31merror^[[0m=runtime error: slice bounds out of range [:100] with capacity 1 ^[[31mstack^[[0m=goroutine 2707009 [running]:^M
runtime/debug.Stack(0xc0000ec000, 0x19b8cb9, 0x5)^M
        GOROOT/src/runtime/debug/stack.go:24 +0x9d^M
github.com/prysmaticlabs/prysm/shared/traceutil.RecoveryHandlerFunc(0x1dc1420, 0xc019088450, 0x18aca20, 0xc009f41ac0, 0xc005dfb1c0, 0x43cc1c)^M
        shared/traceutil/recovery_interceptor_option.go:29 +0x155^M
github.com/grpc-ecosystem/go-grpc-middleware/recovery.recoverFrom(0x1dc1420, 0xc019088450, 0x18aca20, 0xc009f41ac0, 0x1b8ebb8, 0x5f8a00000000, 0x8)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/recovery/interceptors.go:53 +0x5a^M
github.com/grpc-ecosystem/go-grpc-middleware/recovery.UnaryServerInterceptor.func1.1(0x1dc1420, 0xc019088450, 0xc000010138, 0xc005dfbab0)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/recovery/interceptors.go:27 +0x77^M
panic(0x18aca20, 0xc009f41ac0)^M
        GOROOT/src/runtime/panic.go:969 +0x166^M
github.com/prysmaticlabs/prysm/beacon-chain/rpc/beacon.(*Server).ListValidatorBalances(0xc000030ea0, 0x1dc1420, 0xc0190884e0, 0xc014877780, 0xc000030ea0, 0xc00026c4d0, 0x1b90390)^M
        beacon-chain/rpc/beacon/validators.go:167 +0x1755^M
github.com/prysmaticlabs/ethereumapis/eth/v1alpha1._BeaconChain_ListValidatorBalances_Handler.func1(0x1dc1420, 0xc0190884e0, 0x194d160, 0xc014877780, 0x17746a0, 0x2c0ddd8, 0x1900b80, 0x2c0d880)^M
        bazel-out/k8-fastbuild/bin/external/com_github_prysmaticlabs_ethereumapis/eth/v1alpha1/linux_amd64_stripped/go_proto%/github.com/prysmaticlabs/ethereumapis/eth/v1alpha1/beacon_chain.pb.go:4088 +0x89^M
github.com/prysmaticlabs/prysm/beacon-chain/rpc.(*Service).validatorUnaryConnectionInterceptor(0xc00026c280, 0x1dc1420, 0xc0190884e0, 0x194d160, 0xc014877780, 0xc01e114860, 0xc01e114880, 0xc0190884e0, 0x1ddb880, 0x2c0d880, ...)^M
        beacon-chain/rpc/service.go:422 +0x77^M
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1dc1420, 0xc0190884e0, 0x194d160, 0xc014877780, 0x1a157d7, 0x38, 0x1dc1420, 0xc0190884e0)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/chain.go:25 +0x63^M
github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.UnaryServerInterceptor.func1(0x1dc1420, 0xc019088450, 0x194d160, 0xc014877780, 0xc01e114860, 0xc01e1148a0, 0x3, 0x3, 0x1dcdc60, 0xc0170f3bc0)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/tracing/opentracing/server_interceptors.go:31 +0x165^M
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1dc1420, 0xc019088450, 0x194d160, 0xc014877780, 0x38, 0xc001f721e0, 0xc0101c2b30, 0x7f36af9ab701)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/chain.go:25 +0x63^M
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0x1dc1420, 0xc019088450, 0x194d160, 0xc014877780, 0xc01e114860, 0xc01e1148c0, 0xc019088390, 0xc0081d6a68, 0x431246, 0xc01e114900)^M
        external/com_github_grpc_ecosystem_go_grpc_prometheus/server_metrics.go:107 +0xad^M
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1dc1420, 0xc019088450, 0x194d160, 0xc014877780, 0x20, 0xc01e114900, 0x203007, 0x203007)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/chain.go:25 +0x63^M
github.com/grpc-ecosystem/go-grpc-middleware/recovery.UnaryServerInterceptor.func1(0x1dc1420, 0xc019088450, 0x194d160, 0xc014877780, 0xc01e114860, 0xc01e1148e0, 0x0, 0x0, 0x0, 0x0)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/recovery/interceptors.go:31 +0xbc^M
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1dc1420, 0xc019088450, 0x194d160, 0xc014877780, 0xc00edcd200, 0xc01e114800, 0xc0081d6b30, 0x431ab8)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/chain.go:25 +0x63^M
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x1dc1420, 0xc019088450, 0x194d160, 0xc014877780, 0xc01e114860, 0xc01e114880, 0xc0081d6ba0, 0x552c18, 0x1895020, 0xc019088450)^M
        external/com_github_grpc_ecosystem_go_grpc_middleware/chain.go:34 +0xd5^M
```

**Comment by fabdarice** [comment]
I can help w/ @nisdas if you wanna assign this one to me.

**Comment by nisdas** [comment]
Great, thanks ! Assigned to you

**Comment by fabdarice** [comment]
hey, @imnisen I am not able to reproduce what you seem to be experiencing.
I've tried with current `master` your exact HTTP request and the curl is returning 100 balances correctly, even when passing incorrect/unknown public keys (which is also weird, but a different topic).

Trying here also yields the same result: https://api.prylabs.network/#/BeaconChain/ListValidatorBalances

**Comment by fabdarice** [comment]
```
"},{"publicKey":"qp1yDUhM8HoPnrOq8x5uY6uhFilRS5gDYpSlrNLLImdP+bd8AqmHrA9PDW851dSX","index":"233","balance":"31999996507","status":"ACTIVE"},{"publicKey":"pa2MhFZlLulZvmp6wg50fsbc9DSRHZ4rXHP2oqa9yQjQYCXolV0O5Cs4E9fo2+6o","index":"234","balance":"32000053220","status":"ACTIVE"},{"publicKey":"pVGE18Q12KG+CAs3hcDsmcgO+9WFkMS7GutYwyLK8g2VAlhTui4iyIxLJLmD/iro","index":"235","balance":"32000047940","status":"ACTIVE"},{"publicKey":"s8ZMPGoecpAQQdjWu0I6sf5e/CcQtCU6mKyXlBumV3nAhVDwmYZhZjq8ZtGYhqLy","index":"236","balance":"32000049260","status":"ACTIVE"},{"publicKey":"l01spCESs5cePqabmW5oSAyKdyleTFdhAVTbfX8B4gCQz9YCxkh7DVbBPpnO/w35","index":"237","balance":"32000038276","status":"ACTIVE"},{"publicKey":"tUz8HW+UvPPyDtFSeG0CJNPR6YTBNrvZQLdqx1suXwXmkLMBNybKpclakvdzi1CE","index":"238","balance":"32000049260","status":"ACTIVE"},{"publicKey":"sZ2KRbiYtVPiuDRibiDWkzPpjl7JdbnRF65PyNm3jAuWZTQTZ+pAa58npXn4mPps","index":"239","balance":"32000034316","status":"ACTIVE"},{"publicKey":"kJdp65u0lRLbU3zqZjcLPFKYLbDkAqqBDlkpvBG9KyBpN+3JgafJQ+iHE9E84L7m","index":"240","balance":"32000034316","status":"ACTIVE"},{"publicKey":"g+ONBl61DyoHmVGpA47rzS9+GjtWBGigfwe9ey7aXaNyEcwJGM3K9iSg9RiNca6s","index":"241","balance":"32000049260","status":"ACTIVE"},{"publicKey":"l0JbEfOYjyi40cHky4WF/Q3OextpLlfktfWV9D8uPjSctEunQ8SavlOGTS4h6c+u","index":"242","balance":"32000053220","status":"ACTIVE"},{"publicKey":"lY+SqEZCKm2EUpnLf30xbE0eh8xaKqTkQdLT+Ik8yKgEQZukq3Tc8N++nML3Cbio","index":"243","balance":"32000049260","status":"ACTIVE"},{"publicKey":"oT5xCv73FZkDNQbdvpASGlve2BAQtnaF0HI0e6lBeNMFouLFbgnufNP3WhukrJN/","index":"244","balance":"32000049260","status":"ACTIVE"},{"publicKey":"phIuj+TKPq98OmoMJOZie5TFf3PHkeCdmzQh2n5jLjO2PvHo+cvevOVg5DG+HdZ3","index":"245","balance":"32000049260","status":"ACTIVE"},{"publicKey":"gyBHSDBPpouQQSXibPLXFVk+MKO7hrEkL0dIczLokru6KB7JqNXaUPqH3ICogtsz","index":"246","balance":"32000053220","status":"ACTIVE"},{"publicKey":"ggV9WOAzbFsdX2NldUQ07HUfU//J8Z0hhcbs14ZT2I+LMU2lYbqbXYlKyuc4PA2b","index":"247","balance":"32000049260","status":"ACTIVE"},{"publicKey":"jvxlQGKdE9nppK99Gxn3Q6efCaEIxiTT9dPCUYshpmNVVqOndd2h9jraGqd1171Z","index":"248","balance":"32000049260","status":"ACTIVE"},{"publicKey":"tKzIWzeiFCXXpMjaDDBK5Is9sqKJIv8gGV1XKgd7yYAvdtUFGIRDX3PrHWnZAIxe","index":"249","balance":"
```

**Comment by imnisen** [comment]
@fabdarice What's the result of trying with incorrect/unknown public keys?  What crashes my node is a request with unknown pubkeys and page_size as parameters.

When I request with a wrong pubkey:

```
curl -X GET "https://api.prylabs.net/eth/v1alpha1/validators/balances?epoch=32&publicKeys=kjvkS7cV367BceMmxX+EtBYnalDca6GUwOsLZ4Ayj6HV3mZDaFrwbMoMYHaRz/No&pageSize=100" -H "accept: application/json"
```

I don't get any responses. Since I couldn't see the beacon chain node logs of the remote node behind `api.prylabs.net`, I couldn't tell if it crashed.

**Comment by imnisen** [comment]
@fabdarice  I could help fix this if you need.  However, what stops me before is to check other RPCs of the same logic.

**Comment by fab-spring** [comment]
@imnisen feel free! I cannot reproduce it.
I've tried with invalid and incorrect keys.
Even tried on the public prylabs api and local node.
Both would return a page of 100 validators (although weird behavior, should return an error)

**Comment by imnisen** [comment]
@fabdarice Could you describe the "weird behavior", I wonder what's this.

**Comment by 0xKiwi** [comment]
Hey @fabdarice are you still taking care of this issue? Or is it not reproducible?

**Comment by imnisen** [comment]
I have tried with `api.prylabs.ne`
```
curl -i -X GET "https://api.prylabs.net/eth/v1alpha1/validators/balances?epoch=0&pageSize=100&publicKeys=tTq3HTMwewaWqSBotzA85ExuEaFK08dA1eYJHnirROqEB4fhpAjFCgh8mFwBo6Vz" -H "accept: application/json"
HTTP/2 503
content-type: application/grpc
grpc-status: 14
grpc-message: upstream connect error or disconnect/reset before headers. reset reason: connection termination
x-backend: beacon-chain-prod-1
x-source: istio-ingressgateway
content-length: 0
date: Sat, 28 Nov 2020 08:53:58 GMT
server: istio-envoy
x-envoy-upstream-service-time: 278
```

**Comment by fabdarice** [comment]
Hey @0xKiwi ,

I am unable to reproduce what @imnisen is seeing.

See below on my end:
```
curl -i -X GET "https://api.prylabs.net/eth/v1alpha1/validators/balances?epoch=32&publicKeys=
xxxxx&pageSize=100" -H "accept: application/json"
HTTP/2 200
content-type: application/json
x-envoy-upstream-service-time: 5701
x-backend: beacon-chain-prod-2
x-source: unknown
date: Sat, 28 Nov 2020 22:37:41 GMT
server: istio-envoy
x-backend: beacon-chain-prod-2
x-source: istio-ingressgateway

{
 "epoch": "32",
 "balances": [
  {
   "publicKey": "j88oiWqF5edu6eUIQ44j5yU9oaI6ZQHjp9VhglINvPTNtErzJnMYGI8fQWg0IUba",
   "index": "0",
   "balance": "32001143871"
  },
  {
   "publicKey": "hz5z7os+T88dL7DxA2uplqyZELWzSPZDi1+O9QhX1NqQddAhip0bmanq4jWjlwPh",
   "index": "1",
   "balance": "63997944254"
  },
  {
   "publicKey": "jC9TXTvsZflctLpFVWbk7D3o2lwTpoFpng+A15Qtb9y87xjIzxj52hSqN5vdbSnF",
   "index": "2",
   "balance": "31997944254"
  },
  {
   "publicKey": "qNm1tizDEUmtWKKBoik80/TcoRhVyYmD52/7YEedjpjlWS5UFfBACooj792EKzYF",
   "index": "3",
   "balance": "32000843212"
  },
  {
   "publicKey": "rflDJ5Q18cGUrdHN/pnj/eUoTQRRpjgisD7DAbscq0OZ0Bb4EkYfC3GoIGuWyjN4",
   "index": "4",
   "balance": "32000932750"

[.....]
```

**Comment by rauljordan** [comment]
This is now resolved with #8162
